libbpf-tools/cpufreq.c

```diff
@@ -0,0 +1,254 @@
+	int duration;
+	int freq;
+	bool verbose;
+} env = {
+	.duration = -1,
+	.freq = 99,
+};
+
+const char *argp_program_version = "cpufreq 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+const char argp_program_doc[] =
+"Sampling CPU freq system-wide & by process. Ctrl-C to end.\n"
+"\n"
+"USAGE: cpufreq [--help] [-d DURATION] [freq]\n"
+"\n"
+"EXAMPLES:\n"
+"    cpufreq         # sample CPU freq as 99hz (default)\n"
+"    cpufreq -d 5    # sample for 5 seconds only\n"
+"    cpufreq 199     # sample CPU freq as 199hz\n"
+"\n";
+
+static const struct argp_option opts[] = {
+	{ "duration", 'd', "DURATION", 0, "Duration to trace in seconds" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'd':
+		errno = 0;
+		env.duration = strtol(arg, NULL, 10);
+		if (errno || env.duration <= 0) {
+			fprintf(stderr, "Invalid duration: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case ARGP_KEY_ARG:
+		if (pos_args++) {
+			fprintf(stderr,
+				"Unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		errno = 0;
+		env.freq = strtoll(arg, NULL, 10);
+		if (errno || env.freq <= 0) {
+			fprintf(stderr, "Invalid freq (in hz): %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int nr_cpus;
+
+static int open_and_attach_perf_event(int freq, struct bpf_program *prog,
+				struct bpf_link *links[])
+{
+	struct perf_event_attr attr = {
+		.type = PERF_TYPE_SOFTWARE,
+		.freq = 1,
+		.sample_period = freq,
+		.config = PERF_COUNT_SW_CPU_CLOCK,
+	};
+	int i, fd;
+
+	for (i = 0; i < nr_cpus; i++) {
+		fd = syscall(__NR_perf_event_open, &attr, -1, i, -1, 0);
+		if (fd < 0) {
+			fprintf(stderr, "failed to init perf sampling: %s\n",
+				strerror(errno));
+			return -1;
+		}
+		links[i] = bpf_program__attach_perf_event(prog, fd);
+		if (libbpf_get_error(links[i])) {
+			fprintf(stderr, "failed to attach perf event on cpu: "
+				"%d\n", i);
+			links[i] = NULL;
+			close(fd);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+}
+
+static int init_freqs_hmz(__u32 *freqs_mhz, int nr_cpus)
+{
+	char *key = "cpu MHz", *p, line[32];
+	int len = strlen(key), freq_hmz, i;
+	FILE *f;
+
+	f = fopen("/proc/cpuinfo", "r");
+	if (!f)
+		return -1;
+
+	while (fgets(line, sizeof(line), f)) {
+		if (strncmp(line, key, len))
+			continue;
+
+		for (p = line + len; isspace(*p); p++);
+		if (*p != ':')
+			goto err_out;
+		for (++p; isspace(*p); p++);
+		if (!*p)
+			goto err_out;
+		freq_hmz = atoi(p);
+		for (i = 0; i < nr_cpus; i++)
+			freqs_mhz[i] = freq_hmz;
```

Comment:
I don't get this as well. Each CPU has a separate line for cpu MHz, while you are assuming that cpu #0 defines CPU for all other CPUs. I just checked my host, two CPUs definitely have different MHz... So this is just wrong.

And ideally you need to search for `process: %d` first to know which CPU it is. Then search for `cpu MHz: %lf` to get its Mhz. It's quite complicated if done properly, which is why I suggested /sys/devices/system/cpu/cpu*/cpufreq/cpuinfo_cur_freq...

Reply:
Indeed, I was wrong about this. My initial thought was that if the scaling_driver policy has been set, then `tp_btf/cpu_frequency `will be triggered, which means that the initialized value will be likely changed before sampling.  So I only need initial values for constant frequency,  which I think all cpu's freq will be the same, so I only parse cpu #0

But as you said, we can enable a governor on specific cores, so this is just wrong.

I fixed this by parse `/sys/devices/system/cpu/cpu*/cpufreq/scaling_cur_freq` as I found `cpuinfo_cur_freq` may not exist.


> cpuinfo_cur_freq
> Current frequency of the CPUs belonging to this policy as obtained from the hardware (in KHz).
> 
> This is expected to be the frequency the hardware actually runs at. If that frequency cannot be determined, this attribute should not be present.
> 
> scaling_cur_freq
> Current frequency of all of the CPUs belonging to this policy (in kHz).
> 
> In the majority of cases, this is the frequency of the last P-state requested by the scaling driver from the hardware using the scaling interface provided by it, which may or may not reflect the frequency the CPU is actually running at (due to hardware design and other limitations).
> 
> Some architectures (e.g. x86) may attempt to provide information more precisely reflecting the current CPU frequency through this attribute, but that still may not be the exact current CPU frequency as seen by the hardware at the moment.

Reply:
yeah, I found out about it yesterday, and seems like it's what kernel believes the CPU frequency to be, so I think it's ok to stick to scaling_cur_freq